### PR TITLE
Random - Fix usermode thread access

### PR DESCRIPTION
**Pull request by ceolin**
- Add a new API helper to grant access to entropy device
- Remove global entropy device cache

There are fixes necessary to ctr_drbg that I'll address in another pr.

### CODE REVIEW

File: include/random/rand32.h [hunk 1]
```diff
@@ -72,6 +73,31 @@ extern void sys_rand_get(void *dst, size_t len);
...
+
+/**
+ * @brief Allow user thread access to random API.
+ *
+ * Random API that uses entropy device requires that the user thread has
+ * granted access to the entropy driver. This API is a helper that should be
+ * called before a user thread uses any other random API.
+ *
+ * @note This function is not necessary if the thread was already granted
+ * access to the entropy device.
```
Comment (andrewboie) [inline include/random/rand32.h:85]: Maybe mention that this isn't necessary either if the underlying random implementation doesn't use an entropy driver.

File: subsys/random/rand32_entropy_common.c [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <drivers/entropy.h>
+
+void sys_rand_access_grant(struct k_thread *thread)
```
Comment (andrewboie) [inline subsys/random/rand32_entropy_common.c:10]: This function is small enough that I wonder if we could just declare it inline in the header.

**Review by tyrantlizard77 — DISMISSED** [review]

File: subsys/random/rand32_entropy_device.c [hunk 3]
```diff
@@ -9,26 +9,18 @@
...
-		 * if the whole operation isn't atomic.
-		 */
-		dev = device_get_binding(DT_CHOSEN_ZEPHYR_ENTROPY_LABEL);
-		__ASSERT((dev != NULL),
-			"Device driver for %s (DT_CHOSEN_ZEPHYR_ENTROPY_LABEL) not found. "
-			"Check your build configuration!",
-			DT_CHOSEN_ZEPHYR_ENTROPY_LABEL);
-		entropy_driver = dev;
-	}
+	dev = device_get_binding(DT_CHOSEN_ZEPHYR_ENTROPY_LABEL);
```
Comment (andrewboie) [inline subsys/random/rand32_entropy_device.c:19]: hmm this isn't great for performance. device_get_binding() is unfortunately O(n) where n is the number of devices, and I think it does a string compare for each one. It's slow.

we could keep this as a global and put it in `z_libc_partition`...but this isn't ideal from a freedom-from-interference perspective as there's no way currently in Zephyr for different logical applications to get separate memory for library globals.

I suggest leaving `entropy_driver` caching in place if CONFIG_USERSPACE is disabled until we have some better solutions here, as this change will negatively affect non user mode

Comment (ceolin) [reply subsys/random/rand32_entropy_device.c:19]: Agree about performance, that was one of the reason I was reluctant about it.
I need to check with @jukkar what is the use case that he needs it in userspace, but I suspect it is related with moving network stack to userspace, maybe some high-level protocol requires it.
The other solution is make syscalls for these syscalls and then remove syscalls from entropy APIs (that I believe is not necessary).

Comment (jukkar) [reply subsys/random/rand32_entropy_device.c:19]: > but I suspect it is related with moving network stack to userspace, maybe some high-level protocol requires it.

Correct, I need this for example in IPv6 when network stack is running in user mode.

> The other solution is make syscalls for these syscalls and then remove syscalls from entropy APIs (that I believe is not necessary).

Would we need to grant the access if we would make the `sys_rand32_get()` a syscall, isn't that be a simpler for the user of this API too?

Comment (ceolin) [reply subsys/random/rand32_entropy_device.c:19]: yes. Actually looking other implementations like xoroshiro, that I initially though we wouldn't need access to a device we still have problems. All of the rely on globals to save  context.

I really think syscalls here will be much clear than have these globals into partitions that must be explicitly add to user threads memory domains + this api to grant permission to the used device.

**Comment by ceolin** [comment]
#25980 is an alternative implementing syscalls

**Review by d3zd3z — DISMISSED** [review]

**Comment by d3zd3z** [comment]
I don't really have any problem with this change itself, but am a little concerned about the reasons for it. If we are keeping DRBG state in userspace, it makes it a lot more accessible to other threads and attacks. Seeing the internal state of the DRBG is generally a complete loss of security.

**Comment by dleach02** [comment]
I agree with @d3zd3z, I like the syscall version

**Comment by ceolin** [comment]
Closing this one and keeping #25980

**Pull request closed without merging**